pysmt/solvers/bdd.py

```diff
@@ -273,79 +273,52 @@ def walk_bool_constant(self, formula, args):
         else:
             return self.ddmanager.Zero()
 
-    def bdd_to_expr(self, bdd):
-        res = None
-        if bdd == self.ddmanager.Zero():
-            res = self.fmgr.Bool(False)
-        elif bdd == self.ddmanager.One():
-            res = self.fmgr.Bool(True)
-        else:
-            var = self.idx2var[bdd.NodeReadIndex()]
-            t = self.fmgr.Implies(var, self.bdd_to_expr(bdd.T()))
-            e = self.fmgr.Implies(self.fmgr.Not(var), self.bdd_to_expr(bdd.E()))
-            # (v -> t) /\ (!v -> e)
-            res = self.fmgr.And(t, e)
-
-            if bdd.IsComplement():
-                res = self.fmgr.Not(res)
-
-        return res
-
-
-    def bdd_to_expr2(self, bdd):
-        res = None
-        if bdd == self.ddmanager.Zero():
-            res = self.fmgr.Bool(False)
-        elif bdd == self.ddmanager.One():
-            res = self.fmgr.Bool(True)
-        else:
-            var = self.idx2var[bdd.NodeReadIndex()]
-
-            t = self.bdd_to_expr2(bdd.T())
-            e = self.bdd_to_expr2(bdd.E())
-            not_t = self.fmgr.Not(t)
-            not_e = self.fmgr.Not(e)
-
-            if bdd.IsComplement():
-                # (v /\ !t) \/ (!v /\ !e)
-                #     P            N
-                p = self.fmgr.And(var, not_t)
-                n = self.fmgr.And(self.fmgr.Not(var),
-                                  not_e)
-                res = self.fmgr.Or(p, n)
-            else:
-                # (v /\ t) \/ (!v /\ e)
-                #    P             N
-                p = self.fmgr.And(var, t)
-                n = self.fmgr.And(self.fmgr.Not(var), e)
-                res = self.fmgr.Or(p, n)
-
-        return res
-
 
-    def bdd_to_expr3(self, bdd, invert=False):
-        """Convert a bdd into an expression and push negation."""
+    def _walk_back(self, bdd, mgr):
+        stack = [bdd]
 
-        res = None
-        if bdd == self.ddmanager.Zero():
-            res = self.fmgr.Bool(invert)
-        elif bdd == self.ddmanager.One():
-            res = self.fmgr.Bool(not invert)
-        else:
-            var = self.idx2var[bdd.NodeReadIndex()]
-
-            if (bdd.IsComplement() and not invert) or \
-               (not bdd.IsComplement() and invert):
-                ne = self.bdd_to_expr3(bdd.E(), not invert)
-                nt = self.bdd_to_expr3(bdd.T(), not invert)
-                res = self.fmgr.Or(self.fmgr.And(var, nt),
-                                   self.fmgr.And(self.fmgr.Not(var), ne))
-            else:
-                t = self.bdd_to_expr3(bdd.T(), invert)
-                e = self.bdd_to_expr3(bdd.E(), invert)
-                res = self.fmgr.Or(self.fmgr.And(var, t),
-                                   self.fmgr.And(self.fmgr.Not(var), e))
-        return res
+        with LockDdManager(self.ddmanager):
+            while len(stack) > 0:
```

Comment:
Add a comment on the semantics of back_memoization:
- If the node is not there, we haven't explored yet,
- If it is None, we are exploring the children
- Otherwise, it contains the pySMT expressoin